stan/math/rev/mat/fun/adj_jac_apply.hpp

```diff
@@ -120,14 +457,18 @@ struct adj_jac_vari : public vari {
  * variables it defines in the autodiff arena because its destructor will
  * never be called and memory will leak if allocated anywhere else.
  *
+ * Targs (the input argument types) can be any mix of doubles, vars, ints,
+ * std::vectors with double, var, or int scalar components, or
+ * Eigen::Matrix s of any shape with var or double scalar components
+ *
  * @tparam F functor to be connected to the autodiff stack
- * @param x input to the functor
+ * @tparam Targs types of arguments to pass to functor
+ * @param args input to the functor
  * @return the result of the specified operation wrapped up in vars
  */
-template <class F>
-Eigen::Matrix<var, Eigen::Dynamic, 1> adj_jac_apply(
-    const Eigen::Matrix<var, Eigen::Dynamic, 1>& x) {
-  adj_jac_vari<F>* vi = new adj_jac_vari<F>(x);
+template <typename F, typename... Targs>
+Eigen::Matrix<var, Eigen::Dynamic, 1> adj_jac_apply(const Targs&... args) {
+  auto vi = new adj_jac_vari<F, Targs...>(args...);
   Eigen::Matrix<var, Eigen::Dynamic, 1> y(vi->M_);
 
   for (int m = 0; m < y.size(); ++m)
```

Comment:
[optional]
[optional, but a really good idea]

Use `std::copy`:

```
std::copy(&y[0], &y[0] + y.size(), vi->y_vi_);
```

the reason to favor `std::copy` is efficiency:(this from cppreference):

> In practice, implementations of std::copy avoid multiple assignments and use bulk copy functions such as std::memmove if the value type is TriviallyCopyable

And it turns out our `var` are TriviallyCopyable:

```
  EXPECT_TRUE(std::is_trivially_copyable<stan::math::var>::value);
```

passes.